lib/Dialect/SV/Transforms/HWSpecialize.cpp

```diff
@@ -0,0 +1,298 @@
+  assert(parameters.size() != 0);
+  std::string name = moduleOp.getName().str();
+  for (auto param : parameters) {
+    auto paramAttr = param.cast<ParamDeclAttr>();
+    int64_t paramValue = paramAttr.getValue().cast<IntegerAttr>().getInt();
+    name += "_" + paramAttr.getName().str() + "_" + std::to_string(paramValue);
+  }
+
+  // Uniqueness check
+  auto parentModule = moduleOp->getParentOfType<ModuleOp>();
+  int uniqueCntr = 0;
+  std::string uniqueName = name;
+  while (parentModule.lookupSymbol(uniqueName))
+    uniqueName = name + "_" + std::to_string(uniqueCntr++);
+
+  return uniqueName;
+}
+
+// Returns true if any operand or result of 'op' is parametric.
+static bool isParametricOp(Operation *op) {
+  return llvm::any_of(op->getOperandTypes(), isParametricType) ||
+         llvm::any_of(op->getResultTypes(), isParametricType);
+}
+
+// Narrows 'value' using a comb.extract operation to the width of the
+// hw.array-typed 'array'.
+static Value narrowValueToArrayWidth(OpBuilder &builder, Value array,
+                                     Value value) {
+  OpBuilder::InsertionGuard g(builder);
+  if (value.isa<BlockArgument>())
+    builder.setInsertionPointToStart(value.getParentBlock());
+  else
+    builder.setInsertionPointAfter(value.getDefiningOp());
+  auto arrayType = array.getType().cast<hw::ArrayType>();
+  return builder.create<comb::ExtractOp>(
+      value.getLoc(), value, /*lowBit=*/0,
+      llvm::Log2_64_Ceil(arrayType.getSize()));
+}
+
+struct EliminateParamValueOpPattern : public OpRewritePattern<ParamValueOp> {
+  EliminateParamValueOpPattern(MLIRContext *context, ArrayAttr parameters)
+      : OpRewritePattern<ParamValueOp>(context), parameters(parameters) {}
+
+  LogicalResult matchAndRewrite(ParamValueOp op,
+                                PatternRewriter &rewriter) const override {
+    // Substitute the param value op with an evaluated constant operation.
+    auto paramValue =
+        evaluateParametricAttr(op.getLoc(), parameters, op.value());
+    if (failed(paramValue))
+      return failure();
+    rewriter.replaceOpWithNewOp<hw::ConstantOp>(
+        op, op.getType(), paramValue.getValue().getSExtValue());
+    return success();
+  }
+
+  ArrayAttr parameters;
+};
+
+// hw.array_get operations require indexes to be of equal width of the
+// array itself. Since indexes may originate from constants or parameters, emit
+// comb.extract operations to fulfil this invariant.
+struct NarrowArrayGetIndexPattern : public OpConversionPattern<ArrayGetOp> {
+public:
+  using OpConversionPattern<ArrayGetOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ArrayGetOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto inputType = type_cast<ArrayType>(op.input().getType());
+    Type targetIndexType = IntegerType::get(
+        getContext(),
+        inputType.getSize() == 1 ? 1 : llvm::Log2_64_Ceil(inputType.getSize()));
```

Comment:
We really need to support `i0` to avoid this...

Reply:
In lieu of opening that can of worms, we should really add a helper somewhere, maybe in lib/Support, so we can stop sprinkling this throughout the codebase... Not blocking for this PR obviously, but filing an issue since it always comes up: https://github.com/llvm/circt/issues/2909